File: arrays_and_hashing/128_longest_consecutive_sequence.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        """
        Approach:
            - store nums in set to allow for O(1) lookup
            - for each element num:
                - if there exists an element num-1, continue. num is not
                the first element in the sequence
                - set num as the start of a sequence, while next does not
                exist, continue adding to the sequence (Q, optimization: do we need to store
                the values? we just need length)
                - record length of sequence
           - return longest length     
        """
        num_set = set(nums)
        len_lcs = 0
        for num in num_set:
            if num - 1 not in num_set:
                # is start val of sequence
                local_len_cs = 1
                curr_num = num + 1
                while curr_num in num_set:
                    local_len_cs += 1
                    curr_num += 1
                len_lcs = local_len_cs if local_len_cs > len_lcs else len_lcs
        return len_lcs
        

File: arrays_and_hashing/128.longest-consecutive-sequence.cpp
/*
 * @lc app=leetcode id=128 lang=cpp
 *
 * [128] Longest Consecutive Sequence
 */

#include <vector>
#include <set>

using namespace std;

// @lc code=start
class Solution {
public:
    /**
     * Approach (?, O(?)):
     * - iterate list of elements, store in map with value
     * as key and...
     * - Note: you are supposed to use the set datastructure
     * - Iterate every element, store in set
     * - problem: i dont want to iterate in every element, that would
     * balloon time complexity up to O(n^2), traverse only starting elements
     * of a sequence, which we can identify by checking that i-1 does not exist
     * in set
     * - Foreach element,
     *      - traverse the starting element and update the max longest
     *      consecutive sequence
     *     
    */
    int longestConsecutive(vector<int>& nums) {
        int lcs = 0, curr, local_lcs;
        set<int> nums_set;

        for (int num : nums) {
            nums_set.insert(num);
        }

        for (int num : nums_set) {
            // traverse only start element
            local_lcs = 1;
            if (nums_set.find(num - 1) == nums_set.end()) {
                curr = num + 1;
                while (nums_set.find(curr) != nums_set.end()) {
                    curr++;
                    local_lcs++;
                }
            }

            lcs = max(lcs, local_lcs);
        }
        return lcs;
    }
};
// @lc code=end

